File: Recursion/shortest_path_in_matrix/shortest_path_in_matrix.cpp
#include <iostream>
#include <utility>
#include <vector>
#include <algorithm>
#include <numeric>
#include <limits>
using namespace std;

// PROBLEM - Given a matrix, find the shortest path sum from the start to the end of the matrix i.e 0:0 to rows-1:cols-1 of the matrix.

// I solved this myself

// RECURSION IS THE WAY TO SOLVE THIS

// Helper function that will do recursion and return the answer
pair<int, vector<int>> recursion(vector<vector<int>> &matrix, pair<int, int> start, pair<int, int> end, int sum, vector<pair<int, int>> &path, vector<int> &path_values, int rows, int cols, int row, int col) {
    if (make_pair(row, col) == end) {
        return {sum, path_values};
    }

    vector<pair<int, int>> possible_moves;
    if (row+1 < rows && find(path.begin(), path.end(), make_pair(row+1, col)) == path.end()) possible_moves.push_back(make_pair(row+1, col));
    if (row-1 >= 0 && find(path.begin(), path.end(), make_pair(row-1, col)) == path.end()) possible_moves.push_back(make_pair(row-1, col));
    if (col+1 < cols && find(path.begin(), path.end(), make_pair(row, col+1)) == path.end()) possible_moves.push_back(make_pair(row, col+1));
    if (col-1 >= 0 && find(path.begin(), path.end(), make_pair(row, col-1)) == path.end()) possible_moves.push_back(make_pair(row, col-1));

    int shortest_sum = numeric_limits<int>::max();
    vector<int> shortest_path;

    for (auto move : possible_moves) {
        int i = move.first;
        int j = move.second;
        path.push_back(move);
        path_values.push_back(matrix[i][j]);
        pair<int, vector<int>> result = recursion(matrix, start, end, sum + matrix[i][j], path, path_values, rows, cols, i, j);
        path.pop_back();
        path_values.pop_back();

        if (result.first < shortest_sum) {
            shortest_sum = result.first;
            shortest_path = result.second;
        }
    }

    return {shortest_sum, shortest_path};
}

pair<int, vector<int>> find_shortest_route(vector<vector<int>> matrix, pair<int, int>start={-1, -1}, pair<int, int>end={-1, -1}) {
    int rows = matrix.size();
    int cols = matrix[0].size();

    if (rows == 0 || cols == 0) return {0, {}};
    if (start.first == -1 || start.second == -1) start = {0, 0};
    if (end.first == -1 || end.second == -1) end = {rows - 1, cols - 1};

    vector<pair<int, int>> path = {start};
    vector<int> path_values = {matrix[0][0]};
    return recursion(matrix, start, end, matrix[0][0], path, path_values, rows, cols, 0, 0);
}

int main() {
    vector<vector<int>> matrix = {
        {1, 2, 3, 4, 5},
        {6, 7, 8, 9, 1},
        {1, 2, 3, 4, 5},
        {6, 9, 1, 0, 2},
        {7, 5, 6, 1, 8}
    };

    pair<int, vector<int>> ans = find_shortest_route(matrix);
    cout << "Shortest Path Sum: " << ans.first << "\nPath: ";
    for (int val : ans.second) cout << val << " ";
    cout << endl;
}

File: Recursion/shortest_path_in_matrix/shortest_path_in_matrix.py
from typing import List

# PROBLEM - Given a matrix, find the shortest path sum from the start to the end of the matrix i.e 0:0 to rows-1:cols-1 of the matrix.

# I solved this myself

# RECURSION IS THE WAY TO SOLVE THIS

#Defining a helper recursion function that will do the recursion
def recursion(matrix, path, path_values, sum, start: List[int], end : List[int], rows: int, columns: int, row: int=None, col: int=None):

  if not row: row = start[0]
  if not col: col = start[-1]

  if [row, col] == end :
    return [sum, path_values]

  possible_moves = []
  if row+1 < rows and ([row+1, col] != start) and [row+1, col] not in path: possible_moves.append([row+1, col])
  if row-1 >= 0 and ([row-1, col] != start) and [row-1, col] not in path : possible_moves.append([row-1, col])
  if col+1 < columns and ([row, col+1] != start) and [row, col+1] not in path : possible_moves.append([row, col+1])
  if col-1 >= 0 and ([row, col-1] != start) and [row, col-1] not in path : possible_moves.append([row, col-1])

  shortest_sum = float("inf")
  shortest_path = []

  for move in possible_moves:
    path_result = recursion(matrix, path+[move], path_values+[matrix[move[0]][move[1]]], sum+(matrix[move[0]][move[1]]), start, end, rows, columns, move[0], move[1])
    if path_result[0] < shortest_sum:
      shortest_sum = path_result[0]
      shortest_path = path_result[1]

  return [shortest_sum, shortest_path]

def get_shortest_path(matrix, start=None, end=None) :
  if not matrix: return [0, []]
  rows = len(matrix)
  columns = len(matrix[0])

  if not start : start = [0, 0]
  if not end : end = [rows-1, columns-1]

  path = [start]
  path_values = [matrix[start[0]][start[-1]]]

  ans = recursion(matrix, path, path_values, int(matrix[0][0]), start, end, rows, columns)

  return ans


matrix = [
    [1, 2, 3, 4, 5],
    [6, 7, 8, 9, 1],
    [1, 2, 3, 4, 5],
    [6, 9, 1, 0, 2],
    [7, 5, 6, 1, 8]
]

ans = get_shortest_path(matrix)
print(f"Shortest Path Sum : ", ans[0])
print(f"Shortest Path : ", ans[-1])